### DISCUSSION TITLE
Implementation of multi-pages and creation of ViewLocator

### DISCUSSION THREAD

**Discussion by al-zakaria** [AvaloniaUI/Avalonia#14625] [Q&A] [answered]
Hello everyone, I just started studying Avaolonia, I moved here from WPF, if this needs to be indicated, then I’ll immediately say that I use CommunityToolkit. And as soon as I started writing the first application, the question arose of how to make it so that when a button was pressed, a new page with new content would appear, I made my own HomePageView.axaml file, and added the HomePageViewModel.cs class to it (accordingly, I put the first file in the Views folder, and the second one into the ViewModels folder). but even before reaching multi-pages, errors had already appeared, in short, in my MainView I divided it into 2 parts, namely a panel with a ListBox and the main part, they are separated by a Border, and I want to make sure that in the main part all elements are displayed using another page, that is, another page was simply loaded there, but I can’t do it, as it turned out later the project should have a ViewLocaor but I don’t have it, I tried to make it myself and connect it to App.cs but nothing worked for me, in short I don’t know how to do this, how to implement multipage and loading another page, and I don’t know how to make a ViewLocator

here is the link to the project
https://github.com/Zakarayaev/avproj

If anything, I did everything according to the guide from YouTube and of course everything worked out for the author and for some reason his ViewLocator was created automatically along with the project, but I didn’t!

thank you in advance

**Comment by timunie** [comment]
Check out Avalonia.Samples

**Comment by thevortexcloud** [comment]
The view locator is included by default in the MVVM templates. I don't think it's in any of the other templates? If you picked the wrong template you may not have it. Anyway, you don't technically need the view locator. You can manually define data templates just like you can in WPF. But if you want the default view locator the code is here as well as docs on how to use it: https://docs.avaloniaui.net/docs/concepts/view-locator

It also sounds like you want a router. There are plenty of those available or you can make your own fairly easily. The general concept is to have some sort of content control that binds to your page view model, and then have the router update that view model whenever the page needs to change. EG See how ReactiveUI does it: https://docs.avaloniaui.net/docs/concepts/reactiveui/routing

**Reply by al-zakaria** [reply]
I did everything as in the documentation, but I still get a huge number of warnings and errors, and as soon as I add

<Application.DataTemplates>
<local:ViewLocator />
</Application.DataTemplates>

the project stops working again

**Reply by timunie** [reply]
I don't see that ViewLocator.cs file anywhere in your project? Where is it?

**Reply by al-zakaria** [reply]
my mistake, I gave the wrong link
that's the correct one
https://github.com/Zakarayaev/avproj

**Reply by al-zakaria** [reply]
I solved this problem, I actually chose the wrong template! but the problem is that the file is still not visible and in the window it writes to me "Not Found: AvaloniaApplication4.Views.HomePageView"

https://github.com/Zakarayaev/avproj

If anything, I changed the repository, but damn how can I fix it?

**Reply by workgroupengineering** [reply]
The namespace is wrong

https://github.com/Zakarayaev/avproj/blob/08f86e0987c66990e29ebdb32e773345a76375ff/Views/HomePageView.axaml.cs#L5

https://github.com/Zakarayaev/avproj/blob/08f86e0987c66990e29ebdb32e773345a76375ff/Views/HomePageView.axaml#L6

it should be `AvaloniaApplication4.Views`

**Reply by al-zakaria** [reply]
but as soon as I change the namespace name I get an error: The name 'InitializeComponent' does not exist in the current context	AvaloniaApplication4

**Reply by workgroupengineering** [reply]
close e re open vs and build

**Reply by al-zakaria** [reply]
I changed and restarted and rebuilt the project, but still an error, and in addition to the old one, another one appeared

AXN0001	Avalonia x:Name generator was unable to generate names for type 'AvaloniaApplication4.HomePageView'. The type 'AvaloniaApplication4.HomePageView' does not exist in the assembly.

**Reply by thevortexcloud** [reply]
Did you update the namespace in the codebehind and the XAML? They need to match. That error can also happen if you have invalid XAML.

**Reply by al-zakaria** [reply]
I updated everything everywhere, you can do it yourself. because I changed everything in the repository specifically for this purpose. maybe I changed it wrong?

**Reply by thevortexcloud** [reply]
Your class name looks wrong.

https://github.com/Zakarayaev/avproj/blob/a57a7faf3d77b130cdd8a9c0308690cfa905c305/Views/HomePageView.axaml#L8

It should match your code behind class name, just like your MainView does.

**Reply by al-zakaria** [reply]
I changed this but the error remains

**Reply by thevortexcloud** [reply, marked as answer]
I just tried it myself, it builds fine if I fix the class name.

`x:Class="AvaloniaApplication4.Views.HomePageView"`

Is what it should be.

**Reply by al-zakaria** [reply]
it worked! Thank you!
